server/src/main/java/org/spine3/server/aggregate/Aggregate.java

```diff
@@ -248,27 +249,50 @@ private void invokeApplier(Message eventMessage) throws InvocationTargetExceptio
     /**
      * Applies passed events.
      *
-     * <p>The events passed to this method is the aggregate data loaded
-     * by a repository and passed to the aggregate so that it restores its state.
+     * <p>The events passed to this method is the aggregate data (which may include
+     * a {@code Snapshot} loaded by a repository and passed to the aggregate so that
+     * it restores its state.
      *
-     * @param events the list of the events
+     * @param aggregateStateRecord the events to play
      * @throws IllegalStateException if applying events caused an exception, which is set as
      *                               the {@code cause} for the thrown instance
      */
-    void play(Iterable<Event> events) {
+    void play(AggregateStateRecord aggregateStateRecord) {
         createBuilder();
+
+        final Snapshot snapshot = aggregateStateRecord.getSnapshot();
+        if (isNotDefault(snapshot)) {
+            restore(snapshot);
+        }
+
+        final List<Event> events = aggregateStateRecord.getEventList();
+
         try {
             for (Event event : events) {
                 final Message message = getMessage(event);
                 final EventContext context = event.getContext();
                 try {
-                    applyEventOrSnapshot(message);
-                    setVersion(context.getVersion(), context.getTimestamp());
+                    applyEvent(message);
+                    final Version newVersion = context.getVersion();
+                    advanceVersion(newVersion);
                 } catch (InvocationTargetException e) {
                     throw wrappedCause(e);
                 }
             }
         } finally {
+            /*
```

Comment:
This comment block is actually for the `try/finally` block, not the method call.